hpx/lcos/local/latch.hpp

```diff
@@ -106,11 +107,14 @@ namespace hpx { namespace lcos { namespace local
         {
             HPX_ASSERT(n >= 0);
 
-            std::unique_lock<mutex_type> l(mtx_);
-            HPX_ASSERT(counter_ >= n);
+            std::ptrdiff_t new_count = (counter_ -= n);
+            HPX_ASSERT(new_count >= 0);
 
-            if ((counter_ -= n) == 0)
-                cond_.notify_all(std::move(l));    // release the threads
+            if (new_count == 0)
```

Comment:
This change slightly changes the behavior of the latch. Especially when calling `count_up` and `count_down` concurrently, the notification of the condition might happen even though the count was increased in another thread. This might require additional synchronization for the caller. This synchronization might have been needed previously in those use cases anyway.

Reply:
The pre-condition for `count_up` is that the internal counter_ is not zero, i.e. `count_up` is called only before the final thread arrives. This should take care of the situation you described.